presto-main/src/main/java/com/facebook/presto/operator/ChannelSet.java

```diff
@@ -100,16 +108,34 @@ public int size()
             return hash.getGroupCount();
         }
 
-        public void addPage(Page page)
+        /**
+         * This function adds a page to a work object for later processing.
+         * @param page A data page whose data needs to be added to the hash table.
+         * @return The work needs to be done on this page.
+         */
+        public Work<?> addPage(Page page)
         {
             Work<?> work = hash.addPage(page);
-            boolean done = work.process();
-            // TODO: this class does not yield wrt memory limit; enable it
-            verify(done);
+            return work;
+        }
+
+        /**
+         * The call back function to reserve memory.
+         * @return true if the memory reservation was successful.
+         */
+        public boolean updateMemoryReservation()
+        {
+            // If memory is not available, once we return, this operator will be blocked until memory is available.
+            operatorContext.setMemoryReservation(getEstimatedSize());
 
-            if (operatorContext != null) {
-                operatorContext.setMemoryReservation(hash.getEstimatedSize());
```

Comment:
So we no longer update memory reservation when adding a page?

Reply:
We moved the page processing logic to SetBuilderOperator.addInput() and SetBuilderOperator.needsInput(), therefore updating memory reservation are moved to there. This function doesn't do real page processing.